### PR TITLE
support cargo install in our docker environment

### LINKED ISSUE #4099: docker-run.sh uses $HOME/.cargo for CARGO_HOME
#### Problem
Using CARGO_HOME set to HOME/.cargo for crates we want to build and run in CI isn't going to work nicely...   Normally when doing that, we'd add CARGO_HOME/bin to the PATH.  However, were we do that, we're likely to negate the value of using a docker image, which has pre-installed, specific versions of cargo, rust, etc.

#### Proposed Solution
The fix (to enable installing and running any random --bin crate during a build) has two parts:

1. have docker-run.sh not use $HOME/.cargo as CARGO_HOME, have it use the workspace or something else with predictable, controlled contents.
2. update our rust Dockerfiles to put $CARGO_HOME/bin on the PATH via an ENV directive (because using --env from docker-run.sh can't do it correctly)

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
 current docker environment doesn't support "cargo install X && X", which would be nice to have for parts of the build

 using real $HOME/.cargo, CARGO_HOME, and PATH=CARGO_HOME/bin will
  alias/hide rustc, cargo, etc. from the docker image (the whole point
  of the image)

 #### Summary of Changes
 use a workspace-based CARGO_HOME, add it's bin directory to the path for scripts and interactive shells

Fixes #4099

### CODE REVIEW

File: ci/docker-run.sh [hunk 1]
```diff
@@ -41,15 +41,17 @@ ARGS=(
   --rm
 )
 
-if [[ -n $CI ]]; then
-  # Share the real ~/.cargo between docker containers in CI for speed
-  ARGS+=(--volume "$HOME:/home")
```
Comment (mvines) [inline ci/docker-run.sh]: Why was this removed?

Comment (rob-solana) [reply ci/docker-run.sh]: I looked at buildkite-agent's $HOME/.cargo and it had rustc, cargo, etc., which  combined with the BASH_ENV stuff below would mean that buildkite-agent's version of cargo would be run instead of the docker image's version...

Comment (mvines) [reply ci/docker-run.sh]: One of the tricks we use to speed up CI is to share the cache of stuff that gets downloaded into ~/.cargo across jobs.  The lack of this cache is quite visible in the CI run for this PR, check out how the “check” step is much slower than usual and it’s now downloading a bunch of stuff.

Comment (rob-solana) [reply ci/docker-run.sh]: I'm not seeing the download part, is it visible in this PR's CI?  Once a .cargo for a working tree is set up, it should stick (via the .cargo entry in .gitignore), right?

Comment (rob-solana) [reply ci/docker-run.sh]: as an alternative, we could designate another place for CARGO_HOME that's shared by all CI, but that doesn't overlap the user's $HOME/.cargo

as yet another alternative, we could clean up buildkite-agent's ~/.cargo to exclude those things that come with a docker image

Comment (rob-solana) [reply ci/docker-run.sh]: ok, I see the clippy downloads...

Comment (rob-solana) [reply ci/docker-run.sh]: wait...  clippy and fmt are part of the docker image!

Comment (mvines) [reply ci/docker-run.sh]: Hmm..so in general I think we want to not “cargo install” in a rust docker container.  Better to install via the Dockerfile that creates the docker image, into /usr.  I know I know, we haven’t been good about that thus far.   But if we are good about that then we still have ~/.cargo available for the cache of crates that keeps CI running fast(er).

Comment (rob-solana) [reply ci/docker-run.sh]: if we're only running stuff from Dockerfile installations, what will slow down CI?  Registry stuff for the source tree?  Is this a wart on rust or are we doing something wrong?

Comment (mvines) [reply ci/docker-run.sh]: It’s another cargo wart I think.  Cargo home also stores git repos for the dependent crates of solana, can’t seem to separate that into another location from .cargo/bin.  Compare the overall run times of this PR against another recent PR. The test-stable job in this PR is also slower than usual due to a cold cargo home dir.

**Comment by rob-solana** [comment]
closing in favor of #4113
